### ISSUE TITLE
Compass flips north to south.

### ISSUE THREAD

**Issue by csvoisard** [TerriaJS/terriajs#2965]
Steps to reproduce:
1) reset map
2) zoom in a lot
3) middle click top of viewport and drag down (compass flips now)
4) zoom out to confirm flipped compass.

Example using https://nationalmap.gov.au/

![compass-flip-national-map](https://user-images.githubusercontent.com/26971214/39924113-992c2cd0-54f4-11e8-81f7-f1006754e3d1.gif)

**Comment by kring** [comment]
Thanks for the report @csvoisard.  Global Risk Map is running quite an old version of TerriaJS. Are you still able to reproduce this bug in maps running a recent version?  e.g. https://nationalmap.gov.au maybe?

**Comment by csvoisard** [comment]
@kring Yes, the bug works on the national map also.  I replaced the animated GIF in the description.  The new GIF demonstrates the bug using the national map.  You won't have to middle-click drag very much at all to see the compass flip, provided you have zoomed in a lot.

**Comment by kring** [comment]
Thanks, I see the problem too. I think what is happening is that the camera is going a bit past "straight down" and so is actually (slightly) upside-down. As a result, North is behind the camera and TerriaJS shows it at the bottom of the screen. It's not technically wrong (imagine if you kept tipping the camera further in that direction, so that the camera was completely upside down; the compass would be correct in that scenario). But it's confusing. Ideally we wouldn't let the camera tip past vertical. Is this a major cause of concern for you, or is it just a curiosity you happened to notice?

**Comment by csvoisard** [comment]
The severity depends on what you mean by "major"?  Users can work-around by resetting the map, but this issue causes extra clicks/zooms and disrupts user workflow.

**Comment by kring** [comment]
They can also just tip the camera back toward the horizon, right?  It seemed to work in my tests, at least.

**Comment by csvoisard** [comment]
Yes, you are correct! But there is the mismatch between the globe and compass, which becomes evident when the user zooms out.

**Comment by kring** [comment]
Thanks. I didn't mean to dismiss it, I think it's a real issue. I was just assessing how urgently you need a fix.

**Comment by kring** [comment]
I looked into this a bit, an it seems Cesium's `Camera.heading` property is returning an incorrect value in this scenario. I think the problem is that Cesium's camera, when using a `constrainedAxis`, prevents further rotation when the camera position is within `EPSILON2` of being parallel with the constrained axis. (See `rotateVertical` in `Camera.js`). But when computing the heading, Cesium decides that the camera is pointing straight down only if it's within `EPSILON3` of the Z axis. (see `getHeading` in `Camera.js`) Changing the latter epsilon to `EPSILON2` solves the problem, but I'm not sure that won't have other repercussions.

**Comment by csvoisard** [comment]
Changing [`EPSILON3` to `EPSILON2` in getHeading()](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Camera.js#L630) seemed to cause more problems.  The globe flip/flops north and south on _every notch_ when doing mouse wheel zoom in/out!

**Comment by csvoisard** [comment]
This bug does not _depend_ on UI mouse interaction, nor zoom.  The gif below shows how to flip north, south on a "freshly reset" compass, by incrementing the 2nd decimal place of `camera.pitch` value, provided the `camera.heading` and `camera.roll` are also set. 

![camera-setview-magic-orientation](https://user-images.githubusercontent.com/26971214/40510134-b904a79c-5f69-11e8-8239-a18079a66ee2.gif)
